memcached.c

```diff
@@ -1047,207 +880,131 @@ static void conn_set_state(conn *c, enum conn_states state) {
+        resp->item = NULL;
+    }
+    if (resp->write_and_free) {
+        free(resp->write_and_free);
+        resp->write_and_free = NULL;
+    }
+    resp->wbytes = 0;
+    resp->tosend = 0;
+    resp->iovcnt = 0;
+    resp->chunked_data_iov = 0;
+    resp->chunked_total = 0;
+    resp->skip = false;
+}
+
+static void resp_add_iov(mc_resp *resp, const void *buf, int len) {
+    assert(resp->iovcnt < MC_RESP_IOVCOUNT);
+    int x = resp->iovcnt;
+    resp->iov[x].iov_base = (void *)buf;
+    resp->iov[x].iov_len = len;
+    resp->iovcnt++;
+    resp->tosend += len;
+}
+
+// Notes that an IOV should be handled as a chunked item header.
+// TODO: I'm hoping this isn't a permanent abstraction while I learn what the
+// API should be.
+static void resp_add_chunked_iov(mc_resp *resp, const void *buf, int len) {
+    resp->chunked_data_iov = resp->iovcnt;
+    resp->chunked_total = len;
+    resp_add_iov(resp, buf, len);
+}
+
+static bool resp_start(conn *c) {
+    mc_resp *resp = do_cache_alloc(c->thread->resp_cache);
+    //fprintf(stderr, "resp alloc: %d\n", c->thread->resp_cache->total);
+    if (!resp) {
+        return false;
     }
-
-    return 0;
-}
-
-
-/*
- * Adds data to the list of pending data that will be written out to a
- * connection.
- *
- * Returns 0 on success, -1 on out-of-memory.
- * Note: This is a hot path for at least ASCII protocol. While there is
- * redundant code in splitting TCP/UDP handling, any reduction in steps has a
- * large impact for TCP connections.
- */
-
-static int add_iov(conn *c, const void *buf, int len) {
-    struct msghdr *m;
-    int leftover;
-
-    assert(c != NULL);
-
-    if (IS_UDP(c->transport)) {
-        do {
-            m = &c->msglist[c->msgused - 1];
-
-            /*
-             * Limit UDP packets to UDP_MAX_PAYLOAD_SIZE bytes.
-             */
-
-            /* We may need to start a new msghdr if this one is full. */
-            if (m->msg_iovlen == IOV_MAX ||
-                (c->msgbytes >= UDP_MAX_PAYLOAD_SIZE)) {
-                add_msghdr(c);
-                m = &c->msglist[c->msgused - 1];
-            }
-
-            if (ensure_iov_space(c) != 0)
-                return -1;
-
-            /* If the fragment is too big to fit in the datagram, split it up */
-            if (len + c->msgbytes > UDP_MAX_PAYLOAD_SIZE) {
-                leftover = len + c->msgbytes - UDP_MAX_PAYLOAD_SIZE;
-                len -= leftover;
-            } else {
-                leftover = 0;
-            }
-
-            m = &c->msglist[c->msgused - 1];
-            m->msg_iov[m->msg_iovlen].iov_base = (void *)buf;
-            m->msg_iov[m->msg_iovlen].iov_len = len;
-
-            c->msgbytes += len;
-            c->iovused++;
-            m->msg_iovlen++;
-
-            buf = ((char *)buf) + len;
-            len = leftover;
-        } while (leftover > 0);
+    // FIXME: make wbuf indirect or use offsetof or something to zero
+    // this. or a sub struct.
+    memset(resp, 0, 128);
```

Comment:
I'm just going to set this to sizeof(mc_resp) for now. I'd put 128 as a placeholder because I really wanted to memset to an offsetof() but didn't spend the time figuring that out while writing it. probably not worth the optimization for now.